Port from R to python of  Lesson 05 Data visualization with ggplot2

djinnome opened the pull request:
Hi folks,

I just completed my[ instructor training](https://arokem.github.io/2016-12-01-PNNL-ttt/) on Friday, and was looking for how to contribute.   Then I saw the [ggplot2 visualization tutorial in the R ecology lesson](https://github.com/datacarpentry/R-ecology-lesson/blob/gh-pages/05-visualization-ggplot2.Rmd), and thought, "[ggplot has been implemented in python](http://ggplot.yhathq.com/).  Why not just port this whole lesson to python?"
   
I realize this may be more than what was meant in the [how to contribute document](https://github.com/datacarpentry/python-ecology-lesson/blob/gh-pages/CONTRIBUTING.md):
"The plotting lesson could really use some love, and a discussion of exporting high-quality plots."

I also realize that matplotlib is the de facto plotting environment for python, but [ggplot makes it so easy for beginners to make beautiful plots ](http://blog.yhat.com/posts/ggplot-for-python.html) that I thought it would be worth the effort to port the lesson.

Included in this PR are the following files:

- 09-visualization-ggplot-R.ipynb  (This is the original R 05-visualization-ggplot2.md converted to jupyter notebook)
- 09-visualization-ggplot-python.ipynb (This is the port of the R jupyter notebook to Python)
- 09-visualization-ggplot-python.md  (This is the automagically generated markdown file from the Python jupyter notebook so that you don't have to use Jupyter to view the lesson).
- output_*.png  (figures automagically generated when converting ipynb -> md)

wrightaprilm commented:
This looks really nice! I wanted to use this as a point-of-entry to open a larger discussion on the development of graphing lessons in the Python Ecology materials. Traditionally, we've kept the DC materials in Markdown, but I think there are clear advantages to having the plotting exercises in the Jupyter notebook.

1. **Question One**: How do we feel about the use of Jupyter notebook for this lesson? Is having one lesson in a notebook incongruous? 

The next thing here is the use of the ggplot python library. It's open source and currently maintained, but I'm not aware of a maintenance plan in the sort of long-term framework that the matplot people are working with. But it does generally look nicer than matplot. This is also an additional install.

2. **Question Two**: Are we good with the use of this library? 

Lastly, @ jarmokivekas filed an issue recently about where plot images are stored. We've traditionally kept them in the instructor's guide, which might be not intuitive for learners. So I see three options here: either put a disclaimer on the learner page saying where the plots are, make a second version of the materials with the plots embedded, or change nothing.

3. **Question Three**: Pick and option or suggest a new one.

So, I'd like some feedback here from @qjcg, @tracykteal and @djinnome here. Plotting and teaching visualization is a real weakness of mine, so I'm going to beg for some help on making a choice here.

jarmokivekas commented:
I think that with this lesson written in python the [06-plotting-with-matplotlib](https://github.com/datacarpentry/python-ecology-lesson/blob/new-learning-objectives/05-loops-and-functions.md) lesson could pretty much be deprecated. It makes little sense to maintain lessons for two different plotting frameworks. Comparing what is in this lesson and what exists in the maplotlib lesson, the matplotlib one leaves some (quite a lot, even) things to be desired.

This lesson covers everything that's in the current matplotlib lesson, and then some. I like how this is not just a list of how to make basic plots, but it includes justifications an explanation for choosing a visualization that fits the data.

I tend to use `pip` instead of anaconda, since I work with a lot of non-sciency python projects, but `ggplot` is on [this list](https://anaconda.org/bokeh/repo?type=conda&label=main), so I'm under the impression it could be installed with `conda install -c bokeh ggplot`. Installation should't be too much of an issue, don't quote me on that, though.

My apprehension, if any, toward ggplot would be that someone starting out may search for examples and end up finding R example code, and the slight syntax differences between Python and R may be confusing or frustrating.

Personally I think that while matplotlib is definitely the de-facto library in python, it's not like ggplot is some obscure alternative and shouldn't be dismissed on the grounds of it not being matplotlib.

Overall I quite like this lesson :+1:

jarmokivekas commented:
I'll add a couple more quick thoughts:

- I think using pandas is great, and since that uses matplotlib by default, it would be logical to use matplotlib for the plotting lesson as well, although the re-write would require quite a lot of work.
- Matplotlib 2.0 is going to be released soon, it will make the default plot styles look nicer. Favouring ggplot because it has nice defaults would be short-sighted.

As general guidelines I'd suggest that:

- Whenever possible, example plots should have at least axis labels and preferably a title and legend. Figures should be meaningful without having the complete context for it.
- All examples should be done with meaningful data. No plotting `numpy.random.rand()` as function of time.

lexnederbragt commented:
@wrightaprilm On the use of Jupyter Notebook for lessons: Software Carpentry had lessons using the notebook for a while, then deliberately switched to markdown a couple of years ago. I don't remember the exact reason, but I suspect that if you send an email to the SWC discuss list you'll get the story behind it.

(Also, could it be an idea to have the notebook discussion split off from this ggplot discussion in a separate issue?)

Seraf69 commented:
@jarmokivekas You are pointing to lesson 5 not 6.
But I agree with you that ggplot2 is not panacea. Personally, I prefer Matplotlib and looking forward to 2.0
Other experts also have reservations about ggplot2 on the basis of prettiness: http://simplystatistics.org/2016/02/11/why-i-dont-use-ggplot2/

wrightaprilm commented:
@jarmokivekas 

> My apprehension, if any, toward ggplot would be that someone starting out may search for examples > and end up finding R example code, and the slight syntax differences between Python and R may be > confusing or frustrating.

>Personally I think that while matplotlib is definitely the de-facto library in python, it's not like ggplot is >some obscure alternative and shouldn't be dismissed on the grounds of it not being matplotlib.

I wouldn't say that I'm dismissing ggplot on the basis of not being matplotlib. But ggplot is produced by a company, whereas matplot is supported by NumFocus, which supports Data Carpentry, and has a sustainability plan that it perpetually seeks funding for. So the concern there is less that it's not matplotlib, and more that there's not a clear articulation of what happens to this library if yhat stops existing. Who knows how likely that is over the lifetime of the Data Carpentry project and organization, but I think that's worth considering.  

@lexnederbragt 

The main issue is that Jekyll converts .md to the website format, but not .ipynb. In the past, we've insisted on only the MD file being pushed to the repository, but given the integral nature of visualization to this lesson, having the notebook in addition might not be the worst idea. But in some ways that doubles the headaches for maintainers, since the diffs on notebooks can be a little tricky. Making sure both resources (md and notebook) have been updated isn't something I'd really like to have to do on every lesson, unless there's a very good argument for it. But yes, I think I will open a separate issue on it tomorrow.

djinnome commented:
Hi folks,

  Sorry, I was on vacation last week, but I am very happy to see so much discussion on this PR.  

 @wrightaprilm **Question One**: How do we feel about the use of Jupyter notebook for this lesson? Is having one lesson in a notebook incongruous?

Well, there is no need to make an exception for this lesson, which is why I converted the Python Jupyter notebook to regular [markdown](https://github.com/datacarpentry/python-ecology-lesson/pull/96/files#diff-b4b9d1ebab7f4c3ba5fc19b3ff7ed282) using [notedown](https://github.com/aaren/notedown)

With that said, this January, there will be a [Data Carpentry hackathon focused on developing a curriculum for reproducible research using Jupyter](http://www.datacarpentry.org/blog/rr-jupyter/), so in the future, it may be that all data carpentry lessons will be using jupyter notebooks.  And with [notedown](https://github.com/aaren/notedown), it might even be possible to use markdown instead of .ipynb JSON to store the jupyter notebook

Question Two: Are we good with the use of this library?

Well, first of all, `ggplot` is just a thin layer on top of matplotlib that reformulates it's matlab-y way of doing things into a [grammar of graphics](http://link.springer.com/book/10.1007%2F0-387-28695-0) that happens to be as close as possible to R's `ggplot2`.  One could do the exact same thing in pure matplotlib, but it would require **a lot** more code.  In teacher training parlance, `ggplot` reduces the cognitive burden of learning data visualization.

The fact that this lesson is nearly identical to the R ecology lesson also reduces cognitive burden.  The less daylight there is between the two lessons, the easier it will be for our students to learn data carpentry regardless of the language..

With that said, the python  `ggplot` library is not perfect. Some useful features in R `ggplot2` are not implemented in python `ggplot`. For example, because [`geom_point()` and `geom_jitter()` do not work on dtype string](https://github.com/yhat/ggplot/issues/571), to port the following R code:

```
ggplot(data = surveys_complete, aes(x = species_id, y = hindfoot_length)) +
    geom_jitter(alpha = 0.3, color = "tomato") +
  geom_boxplot(alpha = 0) 
```

into python, I had to convert the string dtype to a category dtype and then use the category codes for breaks and the categories as labels:

```
surveys_complete['species_factor'] = surveys_complete['species_id'].astype('category').cat.codes
xlabels = sorted(set(surveys_complete['species_id'].values) )
xcodes = sorted(set(surveys_complete['species_factor'].values))

ggplot(aes(x = 'species_factor', y = 'hindfoot_length'),data = surveys_complete) + \
    geom_jitter(alpha=0.7,jittersize=0.4, color="coral") + \
    scale_x_discrete(breaks=xcodes, labels=xlabels) + \
                         xlab('species_id') + geom_boxplot(alpha=0)
```

Nonetheless, my view is that `ggplot`  is adequate for the specific purpose of teaching students how to quickly get up and running with data visualization in python, which I think is the goal. If they later want to learn a more full-featured python visualization library, such as [bokeh](http://bokeh.pydata.org/en/latest/) or [seaborn](http://seaborn.pydata.org/) or perhaps even [matplotlib 2.0](http://matplotlib.org/devdocs/users/whats_new.html), the grammar of graphics concepts they learned in `ggplot` should serve them well.




I'm not sure I understand **Question 3** but when I converted the jupyter notebook to markdown using [notedown](https://github.com/aaren/notedown), there is an option to strip out all output, so that the plots would only be available in the (unstripped) instructor's guide.

wrightaprilm commented:
OK, so it sounds like you're happy with the compromise of converting to .md. I wanted to throw it out to the community, since my interests as a maintainer are somewhat different than those of contributors and instructors. 

With respect to Q3, we've typically not put the plots directly into the lesson. But I think I'd rather have the plots directly in. It's not really like they can cheat, or that cheating is even meaningful in this context.  So I'm in favor of putting them in the .md. What do you think?

jarmokivekas commented:
I'm also in favour of putting plots is the .md.

There is little difference in the amount of learning done by copy-pasting example code and running it locally versus having the plot already in the .md. Those learners who want to type out the code (which I would encourage) themselves would still be able to do that. 

I think the material is more useful for self-learning and if used as reference material later on if there are plots in the .md

wrightaprilm commented:
The lesson looks good to me. If it looks good to you, @qjcg, I'll go ahead and do the merge, then do some updating on the headers to get @djinnome credit for his contribution.

qjcg commented:
Looks good to me @wrightaprilm, I agree with merging.

djinnome commented:
Yay! Thanks, April.    It looks like there were some minor issues in the conversion between Jupyter and Markdown.  I'll compile a list of them so we can knock them off.
